apptools/persistence/tests/test_state_pickler.py

```diff
@@ -43,69 +46,68 @@
-        self.list = [1, 1.1, "a", 1j, self.inst]
-        self.pure_list = list(range(5))
-        self.dict = {"a": 1, "b": 2, "ref": self.inst}
-        self.numeric = numpy.ones((2, 2, 2), "f")
-        self.ref = self.numeric
+if numpy is not None:
+    from apptools.persistence import state_pickler
+
+    # A simple class to test instances.
+    class A(object):
+
+        def __init__(self):
+            self.a = "a"
+
+    # NOTE: I think that TVTK specific testing should be moved to the
+    #       TVTK package.
+
+    # A classic class for testing the pickler.
+    class TestClassic:
+        def __init__(self):
+            self.b = False
+            self.i = 7
+            self.longi = 1234567890123456789
+            self.f = math.pi
+            self.c = complex(1.01234, 2.3)
+            self.n = None
+            self.s = "String"
+            self.u = "Unicode"
+            self.inst = A()
+            self.tuple = (1, 2, "a", A())
+            self.list = [1, 1.1, "a", 1j, self.inst]
+            self.pure_list = list(range(5))
+            self.dict = {"a": 1, "b": 2, "ref": self.inst}
+            self.numeric = numpy.ones((2, 2, 2), "f")
+            self.ref = self.numeric
+            if TVTK_AVAILABLE:
+                self._tvtk = tvtk.Property()
+
+    # A class with traits for testing the pickler.
+    class TestTraits(HasTraits):
+        b = Bool(False)
+        i = Int(7)
+        longi = Int(12345678901234567890)
+        f = Float(math.pi)
+        c = Complex(complex(1.01234, 2.3))
+        n = Any
+        s = Str("String")
+        u = Str("Unicode")
+        inst = Instance(A)
+        tuple = Tuple
+        list = List
+        pure_list = List(list(range(5)))
+        dict = Dict
+        numeric = Array(value=numpy.ones((2, 2, 2), "f"))
+        ref = Array
         if TVTK_AVAILABLE:
-            self._tvtk = tvtk.Property()
-
-
-# A class with traits for testing the pickler.
-class TestTraits(HasTraits):
-    b = Bool(False)
-    i = Int(7)
-    longi = Int(12345678901234567890)
-    f = Float(math.pi)
-    c = Complex(complex(1.01234, 2.3))
-    n = Any
-    s = Str("String")
-    u = Str("Unicode")
-    inst = Instance(A)
-    tuple = Tuple
-    list = List
-    pure_list = List(list(range(5)))
-    dict = Dict
-    numeric = Array(value=numpy.ones((2, 2, 2), "f"))
-    ref = Array
-    if TVTK_AVAILABLE:
-        _tvtk = Instance(tvtk.Property, ())
-
-    def __init__(self):
-        self.inst = A()
-        self.tuple = (1, 2, "a", A())
-        self.list = [1, 1.1, "a", 1j, self.inst]
-        self.dict = {"a": 1, "b": 2, "ref": self.inst}
-        self.ref = self.numeric
+            _tvtk = Instance(tvtk.Property, ())
+
+        def __init__(self):
+            self.inst = A()
+            self.tuple = (1, 2, "a", A())
+            self.list = [1, 1.1, "a", 1j, self.inst]
+            self.dict = {"a": 1, "b": 2, "ref": self.inst}
+            self.ref = self.numeric
```

Comment:
This was a bit awkward, but these classes should only be defined if we have numpy, so I moved them inside the above if statement.  There is probably a cleaner way to do this

Reply:
Since all of the tests in this module requires NumPy, this is a case where raising `SkipTest` at the module is appropriate, like this: https://github.com/enthought/traitsui/blob/464b20200eef30f23317ffd3f6a125520329f7e2/traitsui/tests/ui_editors/test_data_frame_editor.py#L11-L14

With that, when we run the test with discover, the entire module will be skipped.  And if one wants to run the test module on its own, e.g. `python -m unittest apptools.persistence.tests.test_state_pickler`, they ought to have NumPy because the subject `state_pickler` requires NumPy. It is understandable to get an exception there.

Reply:
I think I flip flop between the two (wrapping code in a condition VS raising SkipTest at the module) depending on the situation (e.g. how optional the dependency is, how much code needs to be wrapped in an if-block, ...). This is an option for consideration.

Reply:
The approach here means in both run scenarios (`unittest discover`) and (`unittest <module-name>`), the tests are consistently skipped in the absence of numpy.

Reply:
In this particular case I went with the approach you described as it is cleaner IMO.  For the rest I left them as is.